### PR TITLE
Add a trailing closure version of GPU device placement API.

### PR DESCRIPTION
**Pull request by rxwei**
Add `withGPU(execute:)`, which takes a closure and return the result that the closure returns (if any).

With this, users can do:
```swift
let x = readTensorFromFile().toAccelerator()
let y = withGPU {
  matmul(x, x).sum(alongAxes: 0, 1)
}
```

This is mechanically the same as declaring a `func` and throwing an `enableGPU()` in it.

Also: 
- Move device placement APIs to `TensorFlow/DevicePlacement.swift`.
- Remove the `infeed` argument from `enableTPU(serverAddress:infeed:)` and make TPU execution always use infeed, since we have standard infeed/outfeed support now.

### CODE REVIEW

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by lattner** [comment]
I'm concerned about this patch, and I don't think syntax is the right direction for this semantic.  The issue to me is that "enableGPU" is a global statement about the nearby tensor code, but this syntax implies strongly that this is a scoped statement about the contents of the closure.  In fact, this:

tensorop
x = withGPU { .. tensorops .. }
tensorop

will end up with all of the tensorops running on the GPU.

I think we should reserve trailing closure syntax like this for a `withDevice(.GPU) { ... }` sort of operation that is equivalent to `tf.device`.

WDYT?

**Comment by mhong** [comment]
Agree that `withDevice(.GPU)` would be better.

On globalness, my understanding is that the closure will become a SIL function, and device info is on a per-function basis: https://github.com/apple/swift/blob/182245a2206666e2cfeab8e6981184b4e10d515a/lib/SILOptimizer/Mandatory/TFDevicePartition.cpp#L57. if the last `tensorop` in the example above is also placed on GPU, that would be a bug that we should fix.

File: test/TensorFlow/device_placement.swift [hunk 1]
```diff
@@ -27,6 +27,16 @@ public func explicitDeviceConfigGPU() {
...
 
+public func closureDeviceConfigGPU() {
+  withGPU {
+    let x = Tensor<Float>(1.0)
+    _hostOp(x)
+  }
+}
+
+// CHECK-LABEL: --- TFPartition Accelerator Result: {{.*}}closureDeviceConfigGPU{{.*}}
+// CHECK: graph_op "Const"() {dtype$dtype: $Builtin.FPIEEE32, value$tensor: f32 0x3F800000 /* 1 */, __device: "ALL_DEVICES"
```
Comment (mhong) [inline test/TensorFlow/device_placement.swift]: scalar promotion always gets onto all devices.

should we extend the test case with something like an add within the closure, and check it's on GPU? Also add some tensor code after the closure, and confirm that's on CPU?

**Comment by rxwei** [comment]
I have no idea why GitHub turned my reply to an edit of an existing comment... Here was my reply:

> The issue to me is that "enableGPU" is a global statement about the nearby tensor code, but this syntax implies strongly that this is a scoped statement about the contents of the closure.

Yes, that’s exactly what I intended to implement here, *not* a closure that senantically imply the placement of sorrounding code.

The implementation is missing something here: I see that this needs a compiler-known symbol similar to TFHoistable.

> I think we should reserve trailing closure syntax like this for a withDevice(.GPU) { ... } sort of operation that is equivalent to tf.device.

I was aware that using an enum to represent the device like `enableDevice(.gpu)` and `withDevice(.gpu) {}` would be the eventual API. However, this patch was trying to minimizing the change with respect to that. 

Now that I see your preference, I'll make the enum change as well.

**Comment by lattner** [comment]
I understand your desire for a global thing, I'm saying that using trailing closure syntax is misleading. I'll drop by to chat.

**Comment by rxwei** [comment]
For the record, we chatted in person and you understood I didn't desire a global thing. It's closure-scoped just as you expected. It's just the implementation was incorrect - the new implementation will be based on emitting start/end points and a CFG walk.

**Comment by lattner** [comment]
+1, sorry for the confusion.  Thanks for tackling this!

**Comment by rxwei** [comment]
Closing because @pschuh took on this.

**Pull request closed without merging**